File: packages/js/src/platform/billing.ts
import type { PlatformClient } from './client'

export interface Balance {
  plan: string
  status: 'active' | 'grace' | 'frozen' | string
  /** Current balance in cents. Negative means the account is in debt. */
  balanceCents: number
  overdraftLimitCents: number | null
  availableCents: number | null
  /** Current burn rate in cents per hour, across every running workload. */
  hourlyRateCents: number
  /** Hours of runway left at the current rate, or `null` when nothing is running. */
  runwayHours: number | null
  expiringCents: number
  expiringAt: number | null
  neverFreeze?: boolean
  invoicedMonthly?: boolean
  email?: string | null
}

export interface Transaction {
  id: string
  kind: string
  amountCents: number
  balanceAfterCents?: number
  description?: string
  createdAt: number
}

export interface TransactionPage {
  items: Transaction[]
  nextCursor: string | null
}

export interface ListTransactionsOpts {
  /** 1-100, default 20. */
  limit?: number
  /** `nextCursor` from a previous page. */
  cursor?: string
  /** Include the daily usage-deduction rows, which are otherwise omitted as noise. */
  includeUsage?: boolean
}

/**
 * Account balance and usage.
 *
 * Billing is **account-scoped, not workspace-scoped**: every workspace you create for
 * a user bills to the account that owns the key. That is what makes per-user
 * workspaces a safe pattern — your users get isolation, you keep one bill — and also
 * what makes {@link Balance.runwayHours} worth watching before you provision more.
 *
 * **Requires an unscoped key.** A scoped key is refused with 403
 * `ACCOUNT_SCOPE_REQUIRED`, because there is no workspace-scoped view of one shared
 * balance, ledger and set of saved cards — and because a scoped key is meant to be
 * handed to an end user, who should not be reading your card details or spending
 * against them. For per-workspace spend, use {@link MetricsAPI.cost} instead.
 */
export class BillingAPI {
  constructor(private readonly client: PlatformClient) {}

  /** Current balance, status, burn rate, and runway. */
  balance(): Promise<Balance> {
    return this.client.get('/api/billing/balance')
  }

  /** A page of balance transactions, newest first. */
  transactions(opts?: ListTransactionsOpts): Promise<TransactionPage> {
    const qs = new URLSearchParams()
    if (opts?.limit !== undefined) qs.set('limit', String(opts.limit))
    if (opts?.cursor) qs.set('cursor', opts.cursor)
    if (opts?.includeUsage) qs.set('includeUsage', '1')
    const q = qs.toString()
    return this.client.get(`/api/billing/transactions${q ? `?${q}` : ''}`)
  }

  /** Cost summary for the current billing period, broken down by resource. */
  summary(): Promise<unknown> {
    return this.client.get('/api/billing/summary')
  }

  /** Live (not-yet-invoiced) usage accumulating right now. */
  live(): Promise<unknown> {
    return this.client.get('/api/billing/live')
  }
}

File: packages/python/lizard/platform/billing.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from .client import PlatformClient


@dataclass
class Balance:
    plan: str
    status: str
    #: Current balance in cents. Negative means the account is in debt.
    balance_cents: int
    #: Current burn rate in cents per hour, across every running workload.
    hourly_rate_cents: int
    #: Hours of runway left at the current rate, or None when nothing is running.
    runway_hours: float | None = None
    overdraft_limit_cents: int | None = None
    available_cents: int | None = None
    expiring_cents: int = 0
    email: str | None = None

    @classmethod
    def _from_dict(cls, d: dict) -> "Balance":
        return cls(
            plan=d.get("plan", ""),
            status=d.get("status", ""),
            balance_cents=int(d.get("balanceCents") or 0),
            hourly_rate_cents=int(d.get("hourlyRateCents") or 0),
            runway_hours=d.get("runwayHours"),
            overdraft_limit_cents=d.get("overdraftLimitCents"),
            available_cents=d.get("availableCents"),
            expiring_cents=int(d.get("expiringCents") or 0),
            email=d.get("email"),
        )


@dataclass
class Transaction:
    id: str
    kind: str
    amount_cents: int
    created_at: int
    balance_after_cents: int | None = None
    description: str | None = None

    @classmethod
    def _from_dict(cls, d: dict) -> "Transaction":
        return cls(
            id=d.get("id", ""),
            kind=d.get("kind", ""),
            amount_cents=int(d.get("amountCents") or 0),
            created_at=int(d.get("createdAt") or 0),
            balance_after_cents=d.get("balanceAfterCents"),
            description=d.get("description"),
        )


@dataclass
class TransactionPage:
    items: list[Transaction] = field(default_factory=list)
    next_cursor: str | None = None


class BillingAPI:
    """
    Account balance and usage.

    Billing is **account-scoped, not workspace-scoped**: every workspace you create
    for a user bills to the account that owns the key. That is what makes per-user
    workspaces a safe pattern -- your users get isolation, you keep one bill -- and
    also what makes :attr:`Balance.runway_hours` worth watching before you
    provision more.
    """

    def __init__(self, client: "PlatformClient") -> None:
        self._client = client

    def balance(self) -> Balance:
        """Current balance, status, burn rate, and runway."""
        return Balance._from_dict(self._client.get("/api/billing/balance"))

    def transactions(
        self,
        *,
        limit: int | None = None,
        cursor: str | None = None,
        include_usage: bool = False,
    ) -> TransactionPage:
        """A page of balance transactions, newest first."""
        params = []
        if limit is not None:
            params.append(f"limit={limit}")
        if cursor:
            params.append(f"cursor={cursor}")
        if include_usage:
            params.append("includeUsage=1")
        qs = f"?{'&'.join(params)}" if params else ""
        body = self._client.get(f"/api/billing/transactions{qs}")
        return TransactionPage(
            items=[Transaction._from_dict(t) for t in (body.get("items") or [])],
            next_cursor=body.get("nextCursor"),
        )

    def summary(self) -> Any:
        """Cost summary for the current billing period, broken down by resource."""
        return self._client.get("/api/billing/summary")

    def live(self) -> Any:
        """Live (not-yet-invoiced) usage accumulating right now."""
        return self._client.get("/api/billing/live")
